### ISSUE TITLE
Extract strings into po files

### ISSUE THREAD

**Issue by purpleidea** [leonelquinteros/gotext#31]
Hey!

I haven't used gettext in a while, and I was looking into a golang variant. Kind of too bad you can't do: `_("the message")` in golang, like you can everywhere else :/

In any case, IIRC (it's been a while) gettext tooling gives you a way to extract all the TO-BE-TRANSLATED strings into the .po files. That way you know what's left to do, and you can update these as you add new strings...

Do you have a script or command that does this? Maybe someone has done some quick golang lexer/parser/token hacking? Not sure how to proceed without it. Thanks!

**Comment by purpleidea** [comment]
Here's something similar which I did not test: https://github.com/gosexy/gettext/tree/master/go-xgettext

**Comment by leonelquinteros** [comment]
Yes, that's something that was brought to my attention before. It's planned, but haven't had the time to implement it.

**Comment by leonelquinteros** [comment]
@purpleidea I've pushed a first approach for extracting translation strings into .po files on master branch. 

Would you like to try it and give some feedback? Please check https://github.com/leonelquinteros/gotext/tree/master/cli/xgotext for more details. 

Note that it's super fresh and isn't feature-complete yet, but it can be a good starting point for not-so-complex cases. 

And of course, ALL feedback is super useful here.

**Comment by MichaelMure** [comment]
This would be super useful, and would probably make me use gotext for https://github.com/MichaelMure/git-bug

**Comment by leonelquinteros** [comment]
@MichaelMure That's great! Cool project BTW. 

It's also useful for me to have adopters so we can test it in the wild and polish some edges. Please send any issues or feature requests you may have and we can improve it together.

**Comment by MichaelMure** [comment]
Well, for now I've looked for solutions to implement i18n for git-bug, but what is available in go is not very convincing. https://github.com/nicksnyder/go-i18n, mentioned in a lot of places, seems to be extremely verbose to use. `gotext` seems quite nice and use the common Po format, but having to extract and manage manually the source strings as for now been enough for me to postpone this task. So, I don't have much feedback apart from that for now, sorry ^^

**Comment by Enrico204** [comment]
@leonelquinteros I'm starting to use `gotext` in my project, and I'm trying the CLI tool (right now). The first thing that I noticed is that the CLI program options are not described in the `README.md`, while the `-help` describe some parameters. As soon as I have some spare time I can make a PR if you want :-)

**Comment by Enrico204** [comment]
and I just found a bug: it's extracting also the first parameter of `grequests.Get()` (which is an URL) :-)

**Comment by leonelquinteros** [comment]
@Enrico204 Thanks for your collaboration! 

PRs are extremely welcome, please send yours whenever you're ready. 

About the bug, yes, it will extract any function signature `Get(a string)` right now, the current implementation is just a proof of concept and needs some progress that I couldn't find time to put into it. You can open a new Issue about that so you can keep track of that progress there and get notified when it's fixed.

**Comment by rodrigo-brito** [comment]
Maybe it is too late, but for future users:
```sh
#!/bin/sh

keywords="GetText"
en_po="etc/locale/en_US/LC_MESSAGES/cifraclub.po"
es_po="etc/locale/es_ES/LC_MESSAGES/cifraclub.po"
temp_en_po="etc/locale/en_US/LC_MESSAGES/cifraclub_temp.po"
temp_es_po="etc/locale/es_ES/LC_MESSAGES/cifraclub_temp.po"

 find . -type f \
	-not -path './vendor/*' \
	-not -path './test/*' \
	-not -name '*_test.go' \
	-name '*.go' \
	| xargs xgettext --keyword="$keywords" --from-code=UTF-8 --language=C++ \
	-o "$1" -

msgmerge $en_po $1 > $temp_en_po
msgmerge $es_po $1 > $temp_es_po
mv $temp_en_po $en_po
mv $temp_es_po $es_po
```

I create a custom function Wrapper to avoid wrong extractions with keyword `Get`.
```go
type GetText struct {
	locale *gotext.Locale
}

 func (g *GetText) GetText(str string, vars ...interface{}) string {
	return g.locale.Get(str, vars...)
}
```
`xgettext` - Searches functions with a given Keyword inside your code. It extracts to a pot file.
`msgmerge` - It will merge a extracted pod file with other po file.

Temp files are required to avoid override of context.

**Comment by leonelquinteros** [comment]
Hey Rodrigo, 

Thanks! This is great. It will be helpful for many users requesting this kind onf solutions. 

With time, I've hope we'll figure out a native way to do this in a clean way, but as you know, you need some spare time to work on open source projects. 

Thanks again for your contribution!

**Comment by bboehmke** [comment]
I tried to improve xgotext in #35.
This should also fix the wrong detection of functions like `grequests.Get()` mentioned by @Enrico204.

Would be great to get some feedback.

**Comment by leonelquinteros** [comment]
@bboehmke 's work has been merged into master branch. 

You're all welcome to heck it, use it and provide feedback in this issue before creating a new release including this new features. 

Thank you all!

**Comment by minjung-ryu** [comment]
The updated patch worked well with my codes. :) 
@leonelquinteros when are you planning to release the new version?

**Comment by bboehmke** [comment]
Sadly I found a couple of issues after running the xgotext on a larger project.

I will try to create a merge request with the fixes in the next days.

**Comment by bboehmke** [comment]
I created a new merge request with the fixes and also an option to exclude directories -> #36

**Comment by leonelquinteros** [comment]
@bboehmke I've merged your changes into master. 

@minjung-ryu Would you like to try it on your side?

**Comment by minjung-ryu** [comment]
@leonelquinteros @bboehmke Yes, I have run extraction and it worked well. And I think extracting to .pot is a good idea :) Thank you for your work.
However I think we don't need to print the.go file paths as log.

**Comment by bboehmke** [comment]
I totally missed this log. I added it because the tool takes over 1 minute for my application and I want to see some progress.

I think the best would be to make it configurable via a flag like "-v" on `go build`.
If I find the time I will add this in the next days.

**Comment by bboehmke** [comment]
Created a merge request #37 to add a verbose flag.
By default nothing is printed and if the "-v" flag is set the tool will print the current directory.

**Comment by leonelquinteros** [comment]
Thanks @bboehmke !!
@minjung-ryu Would you like to check it and provide feedback?

**Comment by minjung-ryu** [comment]
@bboehmke @leonelquinteros Yes, I could configure verbose flag. Thanks for the change :)
And I have one question, does the current version of xgotext support extraction from html template files?

**Comment by chrisvaughn** [comment]
just a note for anyone trying to use the xgotext cli tool on Go 1.14.  It appears there's an issue with x/tools/go/packages in Go 1.14 https://github.com/golang/go/issues/37617 I was getting the same error 'internal error: package "fmt" without types was imported from ...' until I rolled back to Go 1.13.10.

**Comment by leonelquinteros** [comment]
BTW, @minjung-ryu I think there is no support yet for extracting translation strings from HTML template files. 

That may be a longer discussion, as these files aren't Go files, but something else. 
If you want to go into that conversation and work with us to get a solution for it, please let's start a new ticket about that and move forward from there.

**Comment by bboehmke** [comment]
@minjung-ryu Sorry I totally missed the question.
Yes this is possible. This is the reason why I implemented the [`AddParser()`](https://github.com/leonelquinteros/gotext/blob/master/cli/xgotext/parser/parser.go#L16) function. Simply implement a new parser and write your own xgotext CLI tool.

I have used this already in my own application for the extraction of HTML template translation with simple regex. You can find the extractor [here](https://gitlab.com/go4matic/go4matic/-/blob/master/cmd/xgotext/main.go).

@leonelquinteros Maybe we should add a little docu for this or at least create a new issue to keep it in mind.

**Comment by purpleidea** [comment]
I realized you could have a `__` function (two underscores) and I did a little POC:

https://play.golang.org/p/OLsPLgIDX-_w

You can't have a single underscore because it's the special value.

I also realized you can do a dot import:

```
import (
	. "github.com/purpleidea/gettext"
)
```

But sadly this doesn't work with `__` because it's seen as private.

So I made this comment in the hope something can change: https://github.com/golang/go/issues/12750#issuecomment-782567688

Cheers!

**Comment by leonelquinteros** [comment]
Let's consider this issue closed now that we've released `xgotext` in `v1.5.0`.
